File: platform_services.py
"""Small platform boundary for desktop integration.

The rest of the application continues to use Qt widgets, QSettings, and the
existing sync/cache services. This module isolates only operations whose
implementation differs by operating system.
"""
from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path

from PySide6.QtCore import QStandardPaths, QUrl
from PySide6.QtGui import QDesktopServices


APPLICATION_DIRECTORY_NAME = "BaiduPhotoSync"


def app_data_directory() -> Path:
    """Return a writable per-user application directory on every desktop OS."""
    location = QStandardPaths.writableLocation(QStandardPaths.AppDataLocation)
    if location:
        return Path(location)
    # QStandardPaths normally supplies a directory after QApplication is
    # initialized. Keep a safe platform fallback for early-startup code.
    if sys.platform == "win32":
        return Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming")) / APPLICATION_DIRECTORY_NAME
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Application Support" / APPLICATION_DIRECTORY_NAME
    return Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share")) / APPLICATION_DIRECTORY_NAME


def open_system_viewer(path: Path) -> bool:
    """Request the desktop's default associated application for one local file."""
    return QDesktopServices.openUrl(QUrl.fromLocalFile(str(path.resolve())))


def clear_windows_registry_settings() -> None:
    """Remove the legacy QSettings registry key only where the registry exists."""
    if sys.platform != "win32":
        return
    try:
        import winreg

        key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, r"Software\Baidu", 0, winreg.KEY_ALL_ACCESS)
        try:
            winreg.DeleteKey(key, "BaiduPhotoSync")
        finally:
            winreg.CloseKey(key)
    except OSError:
        pass


def remove_application_data() -> None:
    """Delete current and historical per-user application data directories."""
    roots = {app_data_directory()}
    if sys.platform == "win32":
        appdata = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
        roots.update({appdata / "BaiduPhotoSync", appdata / "YikeSync"})
    for root in roots:
        shutil.rmtree(root, ignore_errors=True)


def migrate_legacy_windows_data() -> None:
    """Move an old Windows AppData folder once when the Qt path is empty."""
    if sys.platform != "win32":
        return
    legacy = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming")) / APPLICATION_DIRECTORY_NAME
    current = app_data_directory()
    if legacy == current or not legacy.exists() or current.exists():
        return
    try:
        current.parent.mkdir(parents=True, exist_ok=True)
        shutil.move(str(legacy), str(current))
    except OSError:
        # A failed migration must never prevent startup. Existing runtime logic
        # will create fresh directories as required.
        return

File: src/platform_services.js
"use strict";

/**
 * 平台服务 — 移植自 platform_services.py
 * 用纯 Node.js 实现，无 PySide6/Qt 依赖。
 */
const fs = require("fs");
const path = require("path");
const os = require("os");
const { execSync, spawn } = require("child_process");

const APPLICATION_DIRECTORY_NAME = "BaiduPhotoSync";

/**
 * 返回应用数据目录路径（不创建目录）。
 * 供 migrate 等逻辑在创建目录之前检查路径用。
 */
function _appDataPath() {
  let base;
  if (process.platform === "win32") {
    base = process.env.APPDATA || path.join(os.homedir(), "AppData", "Roaming");
  } else if (process.platform === "darwin") {
    base = path.join(os.homedir(), "Library", "Application Support");
  } else {
    base = process.env.XDG_DATA_HOME || path.join(os.homedir(), ".local", "share");
  }
  return path.join(base, APPLICATION_DIRECTORY_NAME);
}

function app_data_directory() {
  const d = _appDataPath();
  fs.mkdirSync(d, { recursive: true });
  return d;
}

function open_system_viewer(filePath) {
  try {
    let proc;
    if (process.platform === "win32") {
      proc = spawn("cmd", ["/c", "start", "", filePath], { stdio: "ignore", shell: false });
    } else if (process.platform === "darwin") {
      proc = spawn("open", [filePath], { stdio: "ignore", shell: false });
    } else {
      proc = spawn("xdg-open", [filePath], { stdio: "ignore", shell: false });
    }
    // 如果进程成功启动（无 error 事件），视为成功
    return new Promise((resolve) => {
      proc.on("error", () => resolve(false));
      // 给进程 100ms 启动窗口，无 error 即认为启动成功
      setTimeout(() => resolve(true), 100);
    });
  } catch (err) {
    return Promise.resolve(false);
  }
}

function clear_windows_registry_settings() {
  if (process.platform !== "win32") return;
  try {
    execSync(
      `reg delete "HKCU\\Software\\Baidu\\BaiduPhotoSync" /f`,
      { stdio: "ignore" }
    );
  } catch (e) {
    // Key may not exist; ignore
  }
}

function remove_application_data() {
  const roots = new Set([app_data_directory()]);
  if (process.platform === "win32") {
    const appdata = process.env.APPDATA || path.join(os.homedir(), "AppData", "Roaming");
    roots.add(path.join(appdata, "BaiduPhotoSync"));
    roots.add(path.join(appdata, "YikeSync"));
  }
  for (const root of roots) {
    try {
      fs.rmSync(root, { recursive: true, force: true });
    } catch (e) {
      // ignore
    }
  }
}

function migrate_legacy_windows_data() {
  if (process.platform !== "win32") return;
  const legacy = path.join(
    process.env.APPDATA || path.join(os.homedir(), "AppData", "Roaming"),
    APPLICATION_DIRECTORY_NAME
  );
  const current = _appDataPath();   // 不创建目录，仅取路径
  if (legacy === current || !fs.existsSync(legacy) || fs.existsSync(current)) return;
  try {
    fs.mkdirSync(path.dirname(current), { recursive: true });
    fs.renameSync(legacy, current);
  } catch (e) {
    // Failed migration must never prevent startup
  }
}

module.exports = {
  APPLICATION_DIRECTORY_NAME,
  app_data_directory,
  open_system_viewer,
  clear_windows_registry_settings,
  remove_application_data,
  migrate_legacy_windows_data,
};
